isogeny_primes.py

```diff
@@ -793,80 +842,273 @@ def get_aux_primes(K, norm_bound, C_K, h_K, contains_imaginary_quadratic):
+            # means alpha_to_eps_mod_p0 is not in GF(p) so can ignore p and move on
+            return {}
+        possible_values = c_power_12h.nth_root(12 * class_gp_order, all=True)
+        filtered_values = filter_possible_values(possible_values, class_gp_gen,
+                                                 prime_field)
+        output[class_gp_gen] = list(set([x**12 for x in filtered_values]))
+
+    return output
+
+
+def tuple_exp(tup, exp_tup):
+    return tuple((t**e for t, e in zip(tup, exp_tup)))
+
+
+def lifts_in_range(N, res_class):
+
+    output = []
+    p = res_class.modulus()
+    centered_lift = res_class.lift_centered()
+
+    low_run = centered_lift
+
+    while low_run > -N - 0.5:
+        output.append(low_run)
+        low_run = low_run - p
+
+    high_run = centered_lift
+
+    while high_run < N + 0.5:
+        output.append(high_run)
+        high_run = high_run + p
+
+    return list(set(output))
+
+
+def get_prime_gens(C_K, my_gens):
+
+    output = []
+
+    for a_class in my_gens:
+        it = C_K.number_field().primes_of_bounded_norm_iter(800)
+        candidate = next(it)
+        candidate_class = C_K(candidate)
+        while not candidate_class == a_class:
+            candidate = next(it)
+            candidate_class = C_K(candidate)
+        output.append(candidate)
+
+    # while my_gens:
+    #     candidate = next(it)
+    #     candidate_class = C_K(candidate)
+    #     if candidate_class in my_gens:
+    #         output.append(candidate)
+    #         my_gens.remove(candidate_class)
+
+    return output
+
+
+def final_filter(C_K, Kgal, aux_primes, my_gens_ideals, gens_info, p, eps,
+                 embeddings):
+    """The possible isogeny prime p is assumed coprime to the prime ideals in my_gens_ideals
+    at this point."""
+    frak_p0 = Kgal.primes_above(p)[0]  # choice of p_0
+    residue_field = frak_p0.residue_field(names='z')
+    prime_field = GF(p)
+
+    logging.debug(f"Starting final filter for Prime {p} for eps {eps}")
+    logging.debug(f"Kgal: {Kgal}, C_K: {C_K}")
+    logging.debug(f"gen_ideals: {my_gens_ideals}, gen_info: {gens_info}")
+
+    # Step 1
+    possible_vals_at_gens = get_possible_vals_at_gens(gens_info, eps,
+                                                      embeddings,
+                                                      residue_field,
+                                                      prime_field)
+
+    if (not possible_vals_at_gens) or (not all(
+            possible_vals_at_gens.values())):
+        logging.debug(
+            f"Prime {p} for eps {eps} filtered in Step 1 of Heavy filter")
+        logging.debug(f"Possible vals at gens: {possible_vals_at_gens}")
+        return False
+
+    # Step 2
+
+    # a list of tuples
+    possible_vals_cart_prod = list(
+        product(*[possible_vals_at_gens[q] for q in my_gens_ideals]))
+
+    # Step 3
+    for q in aux_primes:
```

Comment:
Currently this final filter is still not as strong as it could be. Every element "possible_vals_cart_prod" represents a character on the galois group. And right now the final filter only stops at step 3b if there is a single aux prime that filters all possible characters at the same time! However it could very well be that each value of "possible_vals_cart_prod" can be ruled out by a different aux prime.

Reply:
i see ... so basically that means that we need another `for` loop?

```
garbage = []
for possible_char in possible_vals_cart_prod:
    for q in aux_primes:
        # essentially the existing code
        if not filtered_values:
            garbage.append(possible_char)
            break
# if not all of possible_vals_cart_prod is garbage, then we still have some possible vals
# so p survives
return len(garbage) == len(possible_vals_cart_prod)
```

Reply:
writing this while connection dropped on call ... for the record, @koffie confirmed that this idea is OK, so i will implement this, test, and if all looks fine, i will merge the PR

Reply:
latest commit has implemented something similar to above. in the end i put the loop over `possible_vals_cart_prod` _inside_ the `aux_primes` loop, because I still think that if `thingy` on the current line 1000 is not in `prime_field`, then we can immediately filter `(p,eps)` at that point; since this is a crucial step, let me know if i'm wrong here

Reply:
No, you are 100% right with your thingy remark.  If thingy is not in prime field, then it can indeed be ruled out, cause then it will never have a 12th root inside prime field, let alone a 12th root that is a frobenius value.

Reply:
great, thanks for confirming